src/Chapter1.hs

```diff
@@ -541,8 +542,15 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+  | x == y && y == z = x
+  | x == y = x
+  | y == z = y
+  | z == x = z
+  | x > y && x < z || x > z && x < y = x
+  | y > x && y < z || y > z && y < x = y
+  | z > x && z < y || z > y && z < x = z
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂 And this otherwise actually can cover the last case as well



Also, It looks like you can move all these `==` cases under other already existing ones by using `>=` and `<=` operators instead of `>` and `<`.

The implementation will be a bit shorter 🙂
```suggestion
  | x >= y && x <= z || x >= z && x <= y = x
  | y >= x && y <= z || y >= z && y <= x = y
  | otherwise = z
```

Reply:
Ah, I see. The last case covers both of the following situations:
1. z > x AND z < y OR vice versa.
2. x, y and z are the same number.
The implementation is now elegant, thanks!